### PR TITLE
Allow longer name for AWS/Azure cluster and refactoring

### PR DESCRIPTION
**Pull request by Michaelvll**
This PR
1. fixes the first part of the TODOs in #963, to unblock the user from using a long name for spot jobs on AWS. 
2. moves the cluster name checking into each Cloud.
3. refactors the `cloud.supports` to `cloud.check_features_are_supported` and change the boolean return value to exception raising.
4. optimizes the failover for `InvalidClusterNameError` and `ClusterUserIdentityError` to directly move to the next cloud, as those errors apply to the whole cloud.
5. fixes minor UX issues for failover:
  Previously:
    image
  Now:
    image

Tested (run the relevant ones):

- [x] Any manual or new tests for this PR (please specify below)
    - [x] `sky launch -c i-am-using-a-very-long-name-for-the-aws-cluster-it-should-be-more-than-63-characters-and-hopefully-it-works-this-sentence-has-135-chars --cloud aws`
    - [x] AWS 244 characters: `sky launch -c i-am-using-a-very-long-name-for-the-aws-cluster-it-should-be-more-than-63-characters-and-hopefully-it-works-by-checking-the-document-of-aws-we-found-that-the-cluster-name-length-should-be-less-than-250-due-to-the-tag-limit-this-sentence-has-246-c --num-nodes 2 --cloud aws -t t3.micro` (and 247 does not work)
    - [x] GCP (didn't change the original tested limit)
    - [x] Azure 42 characters: `sky launch -c i-am-using-a-very-long-name-for-the-aws-cluste --cloud azure --region southcentralus --num-nodes 2`
    - [x]  Lambda 64 characters: `sky launch -c i-am-using-a-very-long-name-for-the-aws-cluster-it-should-be-mor --cloud lambda` works
    - [x] `sky launch -c i-am-using-a-very-long-name-for-the-azure-cluster-it-should-be-more-than-63-characters-and-hopefully-it-works-this-sentence-has-136-chars --cloud azure`: correctly fails for the Azure
    - [x] `sky launch -c i-am-using-a-very-long-name-for-the-aws-cluster-it-should-be-more-than-63-characters-and-hopefully-it-works-this-sentence-has-135-chars --gpus A100:8`, it successfully and quickly failover through Lambda/GCP/Azure and AWS (skip the first three clouds with a single try and try all the regions on AWS).

### CODE REVIEW

File: sky/clouds/cloud.py [hunk 1]
```diff
@@ -66,6 +68,14 @@ class Cloud:
 
     _REPR = '<Cloud>'
 
+    _MAX_CLUSTER_NAME_LEN_LIMIT: Optional[int] = None
+
+    # All features implemented by this cloud. Used by
+    # Cloud.check_features_are_supported()
+    # None means all features are supported.
+    _CLOUD_IMPLEMENTATION_FEATURES: Optional[
+        Set[CloudImplementationFeatures]] = None
```
Comment (concretevitamin) [inline sky/clouds/cloud.py:77]: For discussion: I feel putting these as methods will more clearly remind implementors that they need to be filled in. Here, the naming of the vars suggests they are "constants", which may be neglected?

Comment (concretevitamin) [inline sky/clouds/cloud.py:77]: Should None default to mean none supported? (Or, we can change default val to set().) 

Because we may add new values (for new features) to CloudImplementationFeatures in the future and that doesn't mean existing Clouds that default to None automatically support those.

File: sky/clouds/cloud.py [hunk 2]
```diff
@@ -320,14 +330,55 @@ def need_cleanup_after_preemption(self,
         return False
 
     @classmethod
-    def supports(cls,
-                 requested_features: Set[CloudImplementationFeatures]) -> bool:
-        """Returns whether all of the requested features are supported.
+    def check_features_are_supported(
+            cls, requested_features: Set[CloudImplementationFeatures]):
```
Comment (concretevitamin) [inline sky/clouds/cloud.py:334]: nit: Shall we consistently -> None?

File: sky/clouds/cloud.py [hunk 3]
```diff
@@ -320,14 +330,55 @@ def need_cleanup_after_preemption(self,
...
+        """Errors out if the cloud does not support all requested features.
 
         For instance, Lambda Cloud does not support autostop, so
-        Lambda.support({CloudImplementationFeatures.AUTOSTOP}) returns False.
+        Lambda.support({CloudImplementationFeatures.AUTOSTOP}) raises the
```
Comment (concretevitamin) [inline sky/clouds/cloud.py:338]: need update

File: sky/clouds/cloud.py [hunk 4]
```diff
@@ -320,14 +330,55 @@ def need_cleanup_after_preemption(self,
...
+        if cls._CLOUD_IMPLEMENTATION_FEATURES is None:
+            return
+        unsupported_features = (requested_features -
+                                cls._CLOUD_IMPLEMENTATION_FEATURES)
+        if unsupported_features:
+            unsupported_features_str = ', '.join(
+                [f.name for f in unsupported_features])
+            with ux_utils.print_exception_no_traceback():
+                raise exceptions.NotSupportedError(
+                    f'Cloud {cls._REPR} does not support the following '
```
Comment (concretevitamin) [inline sky/clouds/cloud.py:354]: See another comment?

File: sky/clouds/cloud.py [hunk 5]
```diff
@@ -320,14 +330,55 @@ def need_cleanup_after_preemption(self,
...
+                    f'features: {unsupported_features_str}')
+
+    @classmethod
+    def check_cluster_name_is_valid(cls, cluster_name: str):
```
Comment (concretevitamin) [inline sky/clouds/cloud.py:358]: ditto

File: sky/clouds/cloud.py [hunk 6]
```diff
@@ -320,14 +330,55 @@ def need_cleanup_after_preemption(self,
...
+        """Errors out on invalid cluster names not supported by cloud providers.
+
+        Bans (including but not limited to) names that:
+        - are digits-only
+        - contain underscore (_)
```
Comment (concretevitamin) [inline sky/clouds/cloud.py:363]:
```suggestion
        - contain underscore (_)
        
```

File: sky/clouds/cloud.py [hunk 7]
```diff
@@ -320,14 +330,55 @@ def need_cleanup_after_preemption(self,
...
+            exceptions.InvalidClusterNameError: If the cluster name is invalid.
+        """
+        if cluster_name is None:
+            return
+        valid_regex = '[a-z]([-a-z0-9]*[a-z0-9])?'
+        if re.fullmatch(valid_regex, cluster_name) is None:
+            with ux_utils.print_exception_no_traceback():
+                raise exceptions.InvalidClusterNameError(
+                    f'Cluster name "{cluster_name}" is invalid; '
+                    f'ensure it is fully matched by regex: {valid_regex}')
```
Comment (concretevitamin) [inline sky/clouds/cloud.py:374]:
```suggestion
                    f'ensure it is fully matched by regex (e.g., no underscores or capital letters): {valid_regex}')
```
minor, not feeling strongly: seen such cases

File: sky/cli.py [hunk 8]
```diff
@@ -3077,7 +3077,7 @@ def spot_launch(
     if name is None:
         name = backend_utils.generate_cluster_name()
     else:
-        backend_utils.check_cluster_name_is_valid(name)
+        clouds.Cloud.check_cluster_name_is_valid(name)
```
Comment (concretevitamin) [inline sky/cli.py:3080]: Maybe comment that this does a basic regex check, while name length check will be done when the controller has decided which cloud to use and begins provisioning the spot cluster?

File: sky/backends/cloud_vm_ray_backend.py [hunk 9]
```diff
@@ -1699,6 +1693,17 @@ def provision_with_retries(
                     cluster_exists=cluster_exists)
                 if dryrun:
                     return
+            except (exceptions.InvalidClusterNameError,
+                    exceptions.NotSupportedError,
+                    exceptions.CloudUserIdentityError) as e:
+                # InvalidClusterNameError: cluster name is invalid,
+                # NotSupportedError: cloud does not support requested features,
+                # CloudUserIdentityError: cloud user identity is invalid.
+                # The exceptions above should not be applicable to the whole
```
Comment (concretevitamin) [inline sky/backends/cloud_vm_ray_backend.py:1702]:
```suggestion
                # The exceptions above should be applicable to the whole
```

File: sky/backends/cloud_vm_ray_backend.py [hunk 10]
```diff
@@ -3198,11 +3197,12 @@ def _check_existing_cluster(
...
         usage_lib.messages.usage.set_new_cluster()
         assert len(task.resources) == 1, task.resources
-        resources = list(task.resources)[0]
-        task_cloud = resources.cloud
         # Use the task_cloud, because the cloud in `to_provision` can be changed
         # later during the retry.
-        backend_utils.check_cluster_name_is_valid(cluster_name, task_cloud)
+        resources = list(task.resources)[0]
+        task_cloud = (resources.cloud
+                      if resources.cloud is not None else clouds.Cloud)
```
Comment (concretevitamin) [inline sky/backends/cloud_vm_ray_backend.py:3204]: Would resources.cloud ever be None here?

**Review by concretevitamin — COMMENTED** [review]
Thanks @Michaelvll, and nice tests! Shall we also test a name length close to AWS's char limit?

File: sky/clouds/cloud.py (hunk 7 above)
Comment (Michaelvll) [reply sky/clouds/cloud.py]: Added. Thanks!

File: sky/core.py [hunk 11]
```diff
@@ -312,9 +312,8 @@ def stop(cluster_name: str, purge: bool = False) -> None:
 
     # Check cloud supports stopping instances
     cloud = handle.launched_resources.cloud
-    if not cloud.supports({clouds.CloudImplementationFeatures.STOP}):
-        raise exceptions.NotSupportedError(
-            (f'{cloud} does not support stopping instances.'))
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/core.py:317]: Good point! I changed the implementation by having each cloud setting the unsupported features along with the reason. Wdyt?

File: sky/clouds/cloud.py (hunk 4 above)
Comment (Michaelvll) [reply sky/clouds/cloud.py]: Please check https://github.com/skypilot-org/skypilot/pull/1648#discussion_r1091529257

File: sky/clouds/cloud.py (hunk 5 above)
Comment (Michaelvll) [reply sky/clouds/cloud.py:379]: Sorry, what does this mean?

**Comment by Michaelvll** [comment]
> Shall we also test a name length close to AWS's char limit?

Good catch! Actually, we have to reduce the limit to 246 for AWS because of the DescribeInstance API.

File: sky/clouds/aws.py [hunk 12]
```diff
@@ -65,6 +66,15 @@ class AWS(clouds.Cloud):
         'https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-quickstart.html'  # pylint: disable=line-too-long
     )
 
+    @classmethod
+    def _cloud_unsupported_features(
+            cls) -> Dict[clouds.CloudImplementationFeatures, str]:
+        return dict()
+
+    @classmethod
+    def _max_cluster_name_len_limit(cls) -> Optional[int]:
```
Comment (concretevitamin) [inline sky/clouds/aws.py:75]:
```suggestion
    def _max_cluster_name_length(cls) -> Optional[int]:
```

File: sky/clouds/cloud.py (hunk 5 above)
Comment (concretevitamin) [reply sky/clouds/cloud.py:379]: Can we add `-> None`?

File: sky/core.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (concretevitamin) [reply sky/core.py:317]: Can we add a note to CloudImplementationFeatures that if any value (feature) is added to that enum, audit all Cloud implementations and potentially add that feature to _cloud_unsupported_features()?

This is related to another comment. However, I do see that to provide a reason message for each unsupported feature this current approach seems appropriate.

File: sky/clouds/lambda_cloud.py [hunk 13]
```diff
@@ -16,18 +16,36 @@
...
-# for Lambda Cloud
-_LAMBDA_IMPLEMENTATION_FEATURES: Set[clouds.CloudImplementationFeatures] = set()
-
 
 @clouds.CLOUD_REGISTRY.register
 class Lambda(clouds.Cloud):
     """Lambda Labs GPU Cloud."""
 
     _REPR = 'Lambda'
+    # Lamdba has a 64 char limit for cluster name.
```
Comment (concretevitamin) [inline sky/clouds/lambda_cloud.py:25]: nit: a blank line between definitions?

File: sky/clouds/lambda_cloud.py [hunk 14]
```diff
@@ -16,18 +16,36 @@
...
+    # Reference: https://cloud.lambdalabs.com/api/v1/docs#operation/launchInstance # pylint: disable=line-too-long
+    _MAX_CLUSTER_NAME_LEN_LIMIT = 64
```
Comment (concretevitamin) [inline sky/clouds/lambda_cloud.py:27]: For all such Cloud's own limits, don't the `ray-` prefix and suffix occupy some chars? Can we test close to the limit for all 4 clouds?

File: sky/clouds/cloud.py [hunk 15]
```diff
@@ -320,14 +348,62 @@ def need_cleanup_after_preemption(self,
...
-    def supports(cls,
-                 requested_features: Set[CloudImplementationFeatures]) -> bool:
-        """Returns whether all of the requested features are supported.
+    def check_features_are_supported(
+            cls, requested_features: Set[CloudImplementationFeatures]) -> None:
+        """Errors out if the cloud does not support all requested features.
 
         For instance, Lambda Cloud does not support autostop, so
-        Lambda.support({CloudImplementationFeatures.AUTOSTOP}) returns False.
+        Lambda.check_features_are_supported({
```
Comment (concretevitamin) [inline sky/clouds/cloud.py:356]: func name needs update

File: sky/clouds/lambda_cloud.py [hunk 16]
```diff
@@ -16,18 +16,36 @@
...
+    # Currently, none of clouds.CloudImplementationFeatures are implemented
+    # for Lambda Cloud.
+    # STOP/AUTOSTOP: The Lambda cloud provider does not support stopping VMs.
+    # MULTI_NODE: Multi-node is not supported by the implementation yet.
+    _CLOUD_UNSUPPORTED_FEATURES = {
+        clouds.CloudImplementationFeatures.STOP: 'Lambda cloud does not support stopping VMs.',
+        clouds.CloudImplementationFeatures.AUTOSTOP: 'Lambda cloud does not support stopping VMs.',
+        clouds.CloudImplementationFeatures.MULTI_NODE: 'Multi-node is not supported by the implementation yet.',
```
Comment (concretevitamin) [inline sky/clouds/lambda_cloud.py:35]:
```suggestion
        clouds.CloudImplementationFeatures.MULTI_NODE: 'Multi-node is not supported by the Lambda Cloud implementation yet.',
```

File: sky/clouds/cloud.py [hunk 17]
```diff
@@ -66,6 +69,31 @@ class Cloud:
...
+            A dict of {feature: reason} for the features not supported by the
+            cloud implementation.
+        """
+        raise NotImplementedError
+
+    @classmethod
+    def _max_cluster_name_len_limit(cls) -> Optional[int]:
+        """Returns the maximum length limit of a cluster name.
+
+        This method is used by check_cluster_name() to check if the cluster
```
Comment (concretevitamin) [inline sky/clouds/cloud.py:90]:
```suggestion
        This method is used by check_cluster_name_is_valid() to check if the cluster
```

**Review by concretevitamin — APPROVED** [review]
Thanks @Michaelvll. Just some minor comments and a question about some additional characters added by the autoscaler. May be worth updating PR description with each cloud's at-limit test?

File: sky/core.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/core.py:317]: Good point! Added. Thanks!

**Comment by Michaelvll** [comment]
> May be worth updating PR description with each cloud's at-limit test?

Great point! I added the tests for each cloud, and seems there are multiple corner cases that are caught by those tests. I fixed the cluster name limit for AWS and Azure. Thanks!
